This notebook is for the captsone project in a 9-course series from the IBM Data Science course on Coursera, which combines material learned from areas in data preprocessing, data analysis, data visualization, and machine learning. 

In [1]:
print("Hello Capstone Project Course!")

Hello Capstone Project Course!


In this assignment, you will be required to explore, segment, and cluster the neighborhoods in the city of Toronto. However, unlike New York, the neighborhood data is not readily available on the internet. What is interesting about the field of data science is that each project can be challenging in its unique way, so you need to learn to be agile and refine the skill to learn new libraries and tools quickly depending on the project.

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. You will be required to scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas dataframe so that it is in a structured format like the New York dataset.

Once the data is in a structured format, you can replicate the analysis that we did to the New York City dataset to explore and cluster the neighborhoods in the city of Toronto.

In [2]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Libraries imported.


In [3]:
pip install lxml

Note: you may need to restart the kernel to use updated packages.


In [4]:
columns = ['PostalCode', 'Borough', 'Neighborhood']

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=columns)

In [5]:
#Read Wikipedia table into pandas dataframe
link= 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
neighborhoods = pd.read_html(link)

In [6]:
#Convert html table to pandas dataframe
neighborhoods = neighborhoods[0]
neighborhoods

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M5A  Downtown Toronto   
6        M6A        North York   
7        M6A        North York   
8        M7A      Queen's Park   
9        M8A      Not assigned   
10       M9A         Etobicoke   
11       M1B       Scarborough   
12       M1B       Scarborough   
13       M2B      Not assigned   
14       M3B        North York   
15       M4B         East York   
16       M4B         East York   
17       M5B  Downtown Toronto   
18       M5B  Downtown Toronto   
19       M6B        North York   
20       M7B      Not assigned   
21       M8B      Not assigned   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M9B         Etobicoke   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M1C       Scarborough   
30       M2C      Not assigned   
31       M3C        North York   
32       M3C        North York   
33       M4C         East York   
34       M5C  Downtown Toronto   
35       M6C              York   
36       M7C      Not assigned   
37       M8C      Not assigned   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M9C         Etobicoke   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M1E       Scarborough   
45       M2E      Not assigned   
46       M3E      Not assigned   
47       M4E      East Toronto   
48       M5E  Downtown Toronto   
49       M6E              York   
50       M7E      Not assigned   
51       M8E      Not assigned   
52       M9E      Not assigned   
53       M1G       Scarborough   
54       M2G      Not assigned   
55       M3G      Not assigned   
56       M4G         East York   
57       M5G  Downtown Toronto   
58       M6G  Downtown Toronto   
59       M7G      Not assigned   
60       M8G      Not assigned   
61       M9G      Not assigned   
62       M1H       Scarborough   
63       M2H        North York   
64       M3H        North York   
65       M3H        North York   
66       M3H        North York   
67       M4H         East York   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M5H  Downtown Toronto   
71       M6H      West Toronto   
72       M6H      West Toronto   
73       M7H      Not assigned   
74       M8H      Not assigned   
75       M9H      Not assigned   
76       M1J       Scarborough   
77       M2J        North York   
78       M2J        North York   
79       M2J        North York   
80       M3J        North York   
81       M3J        North York   
82       M4J         East York   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M5J  Downtown Toronto   
86       M6J      West Toronto   
87       M6J      West Toronto   
88       M7J      Not assigned   
89       M8J      Not assigned   
90       M9J      Not assigned   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M1K       Scarborough   
94       M2K        North York   
95       M3K        North York   
96       M3K        North York   
97       M4K      East Toronto   
98       M4K      East Toronto   
99       M5K  Downtown Toronto   
100      M5K  Downtown Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M6K      West Toronto   
104      M7K      Not assigned   
105      M8K      Not assigned   
106      M9K      Not assigned   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M1L       Scarborough   
110      M2L        North York   
111      M2L        North York   
112      M3L        North York   
113      M4L      East Toronto   
114      M4L      East Toronto   
115      M5L  Downtown Toronto   
116      M5L  Downtown Tor

In [7]:
neighborhoods.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [8]:
neighborhoods.shape

(288, 3)

Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

In [9]:
#Delete rows with unassigned Boroughs
indexNames = neighborhoods[neighborhoods['Borough'] == 'Not assigned' ].index

# Delete these row indexes from dataFrame
neighborhoods.drop(indexNames , inplace=True)

More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.

In [10]:
#Use groupby to merge rows with matching postcodes and combine neighbourhoods
df = neighborhoods.groupby(by=['Postcode','Borough']).agg(lambda x: ', '.join(x))
df.reset_index(level=['Postcode','Borough'], inplace=True)

In [11]:
df.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.

In [12]:
#Replace any instance of 'Not assigned' in the Neighbourhood column with the value from the Borough column for that row
df['Neighbourhood'] = np.where(df['Neighbourhood'] == 'Not assigned', df['Borough'], df['Neighbourhood'])

Print out the shape of our dataframe df to show that all data preprocessing has been done successfully and we have now the data ready to be analyzed. 

(number of instances should be less that original count of 288)

In [13]:
df.shape

(103, 3)

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In [14]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [15]:
import geocoder

In [16]:
#convert postcode series to list to iterate through
pc_list = df['Postcode'].tolist()

#two lists to store all lat and lon values
lat_vals = []
lon_vals = []

#loop through all values in pc_list to get lat and lon values for each item
for code in pc_list:

    # initialize your variable to None
    lat_lng_coords = None

    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(code))
        lat_lng_coords = g.latlng
    #add lat and lon values for code to list
    lat_vals.append(lat_lng_coords[0])
    lon_vals.append(lat_lng_coords[1])
    
#add each list to pandas dataframe
df['Latitude'] = lat_vals
df['Longitude'] = lon_vals

df.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.811525   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.785730   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.765690   
3      M1G  Scarborough                                  Woburn  43.768359   
4      M1H  Scarborough                               Cedarbrae  43.769688   

   Longitude  
0 -79.195517  
1 -79.158750  
2 -79.175256  
3 -79.217590  
4 -79.239440

Now that I have all of the appropriate latitudes and longitudes for each postal code, I am going to create a new dataframe to segment and cluster the neighborhoods in the Boroughs of Toronto.

In [17]:
#create a new dataframe with all Boroughs containing "Toronto"
toronto_df = df[df['Borough'].str.contains("Toronto")].reset_index(drop=True)
toronto_df.head()

Postcode          Borough                   Neighbourhood   Latitude  \
0      M4E     East Toronto                     The Beaches  43.676845   
1      M4K     East Toronto    The Danforth West, Riverdale  43.683262   
2      M4L     East Toronto  The Beaches West, India Bazaar  43.667965   
3      M4M     East Toronto                 Studio District  43.662766   
4      M4N  Central Toronto                   Lawrence Park  43.728160   

   Longitude  
0 -79.295225  
1 -79.355120  
2 -79.314673  
3 -79.334830  
4 -79.387085

In [18]:
toronto_df.shape

(38, 5)

In [19]:
#Get the coordinates for Toronto to map
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="t_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [20]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [21]:
#create necessary variables from Foursquare API credentials
CLIENT_ID = 'YTP01X4D1JAYDFMR1QEBL04OY1TWI0LJRBXACDZKVARDICCU' # your Foursquare ID
CLIENT_SECRET = 'OLPDOI3EVXFTSLQB2XUDT4W0ONHZMEMPMG0R0YKAMYCKGWIL' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: YTP01X4D1JAYDFMR1QEBL04OY1TWI0LJRBXACDZKVARDICCU
CLIENT_SECRET:OLPDOI3EVXFTSLQB2XUDT4W0ONHZMEMPMG0R0YKAMYCKGWIL


In [22]:
#create a function to explore all venues nearby each postal code of our new Toronto dataframe
LIMIT = 500
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [23]:
#Run the function on each neighborhood to get nearby venues
toronto_venues = getNearbyVenues(names=toronto_df['Neighbourhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

In [24]:
print(toronto_venues.shape)
toronto_venues.head()

(1744, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676845   
1                   The Beaches              43.676845   
2                   The Beaches              43.676845   
3                   The Beaches              43.676845   
4  The Danforth West, Riverdale              43.683262   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.295225                   Glen Manor Ravine       43.676821   
1              -79.295225  The Big Carrot Natural Food Market       43.678879   
2              -79.295225                 Grover Pub and Grub       43.679181   
3              -79.295225                       Upper Beaches       43.680563   
4              -79.355120                           Dollarama       43.686197   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.355989     Discount Store

In [25]:
#check how many venues were returned for each neighborhood
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            64   
Brockton, Exhibition Place, Parkdale Village                           66   
Business Reply Mail Processing Centre 969 Eastern                     100   
CN Tower, Bathurst Quay, Island airport, Harbou...                     67   
Cabbagetown, St. James Town                                            40   
Central Bay Street                                                     97   
Chinatown, Grange Park, Kensington Market                              95   
Christie                                                               10   
Church and Wellesley                                                   85   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             26   
Davisville North                                                        9   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                      7   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           19   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     3   
Harbord, University of Toronto                                         52   
Harbourfront East, Toronto Islands, Union Station                       4   
Harbourfront, Regent Park                                              26   
High Park, The Junction South                                           1   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               63   
Moore Park, Summerhill East                                             3   
North Toronto West                                                      4   
Parkdale, Roncesvalles                                                 49   
Rosedale                                                                4   
Runnymede, Swansea                                                     41   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                       100   
Studio District                                                        54   
The Annex, North Midtown, Yorkville                                    27   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         17   
The Danforth West, Riverdale                                            4   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             64   
Brockton, Exhibition Place, Parkdale Village                            66   
Business Reply Mail Processing Centre 969 Eastern                      100   
CN Tower, Bathurst Quay, Island airport, Harbou...                      67   
Cabbagetown, St. James Town                                             40   
Central Bay Street                                                      97   
Chinatown, Grange Park, Kensington Market                               95   
Christie                                                                10   
Church and Wellesley                                                    85   
Commerce Court, Victoria Hotel                                         100   
Davisville                        

Now I will use one hot encoding to prepare the data to be run with kMeans

In [26]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  American Restaurant  Antique Shop  \
0            0                  0                    0             0   
1            0                  0                    0             0   
2            0                  0                    0             0   
3            0                  0                    0             0   
4            0                  0                    0             0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 0   

   Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                   0          0           0           0       0     0    0   
1                   0          0           0           0       0     0    0   
2                   0          0           0           0       0     0    0   
3                   0          0           0           0       0     0    0   
4                   0          0           0           0       0     0    0   

   Baseball Field  Basketball Stadium  Beer Bar  Beer Store  \
0               0                   0         0           0   
1               0                   0         0           0   
2               0                   0         0           0   
3               0                   0         0           0   
4               0                   0         0           0   

   Belgian Restaurant  Bistro  Bookstore  Boutique  Brazilian Restaurant  \
0                   0       0          0         0                     0   
1                   0       0          0         0                     0   
2                   0       0          0         0                     0   
3                   0       0          0         0                     0   
4                   0       0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Camera Store  Caribbean Restaurant  \
0              0         0        0     0             0                     0   
1              0         0        0     0             0                     0   
2              0         0        0     0             0                     0   
3              0         0        0     0             0                     0   
4              0         0        0     0             0                     0   

   Cheese Shop  Chinese Restaurant  Church  Clothing Store  Cocktail Bar  \
0            0                   0       0               0             0   
1            0                   0       0               0             0   
2            0                   0       0               0             0   
3            0                   0       0               0             0   
4            0                   0       0               0             0   

   Coffee Shop  College Arts Building  College Gym  College Rec Center  \
0            0                      0            0                   0   
1            0                      0            0                   0   
2            0                      0            0                   0   
3            0                      0            0                   0   
4            0                      0            0                   0   

   Colombian Restaurant  Com

In [27]:
#check the size of the resulting dataframe
toronto_onehot.shape

(1744, 213)

In [28]:
#Group rows by neighborhood and take the mean of the frequency of occurrence of each category
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.000000   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.014925   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.000000   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.000000   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.019231   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.038462   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.015873   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.000000   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                 Runnymede, Swansea     0.000000   
29                           Ryerson, Garden District     0.000000   
30                                     St. James Town     0.000000   
31                    Stn A PO Boxes 25 The Esplanade     0.000000   
32                                    Studio District     0.000000   
33                The Annex, North Midtown, Yorkville     0.000000   
34                                        The Beaches     0.000000   
35                     The Beaches West, India Bazaar     0.000000   
36                       The Danforth West, Riverdale     0.000000   

    Afghan Restaurant  American Restaurant  Antique Shop  Art Gallery  \
0            0.000000             0.030000      0.000000     0.010000   
1            0.000000             0.000000      0.000000     0.015625   
2            0.000000             0.000000      0.000000     0.015152   
3            0.000000             0.030000      0.000000     0.000000   
4            0.000000             0.000000      0.000000     0.000000   
5            0.000000             0.000000      0.000000     0.000000   
6            0.000000             0.010309      0.000000     0.010309   
7            0.000000             0.000000      0.000000     0.010526   
8            0.000000             0.000000      0.000000     0.000000   
9            0.011765             0.011765      0.000000     0.000000   
10           0.000000             0.040000      0.000000     0.010000   
11           0.000000             0.038462      0.000000     0.000000   
12           0.000000             0.111111      0.000000     0.000000   
13           0.000000             0.000000      0.000000     0.000000   
14           0.000000             0.030000      0.000000     0.010000   
15           0.000000             0.000000      0.000000     0.052632   
16           0.000000             0.030000      0.000000     0.010000   
17           0.000000             0.000000      0.000000     0.000

In [29]:
toronto_grouped.shape

(37, 213)

Print each neighborhood along with the top 5 venues associated

In [30]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.09
1                 Café  0.06
2                Hotel  0.04
3                  Bar  0.03
4  Japanese Restaurant  0.03


----Berczy Park----
            venue  freq
0     Coffee Shop  0.08
1    Cocktail Bar  0.05
2      Restaurant  0.05
3          Bakery  0.03
4  Breakfast Spot  0.03


----Brockton, Exhibition Place, Parkdale Village----
                    venue  freq
0             Coffee Shop  0.11
1                    Café  0.06
2  Furniture / Home Store  0.05
3                  Bakery  0.05
4              Restaurant  0.05


----Business Reply Mail Processing Centre 969 Eastern----
         venue  freq
0  Coffee Shop  0.09
1         Café  0.05
2          Bar  0.04
3   Steakhouse  0.04
4        Hotel  0.04


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
                venue  freq
0         Coffee Shop  0.09
1  Italian Restaurant

Function to sort values in ascending order

In [31]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create the new dataframe and display the top 10 venues for each neighborhood

In [32]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern           Coffee Shop   
4  CN Tower, Bathurst Quay, Island airport, Harbo...           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
0                  Café                 Hotel        Asian Restaurant   
1            Restaurant          Cocktail Bar                Beer Bar   
2                  Café            Restaurant  Furniture / Home Store   
3                  Café                 Hotel              Steakhouse   
4    Italian Restaurant            Restaurant                    Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse    Seafood Restaurant   American Restaurant   
1                Bakery        Breakfast Spot                 Hotel   
2                Bakery        Sandwich Place              Beer Bar   
3                   Bar      Sushi Restaurant   American Restaurant   
4                   Bar                  Park                   Pub   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Japanese Restaurant          Burger Joint              Gastropub  
1                  Café                Lounge            Cheese Shop  
2           Supermarket                   Bar     Italian Restaurant  
3   Japanese Restaurant    Italian Restaurant            Pizza Place  
4             Speakeasy                Bakery         Sandwich Place

Run kMeans to cluster each neighborhood into 4 different clusters

In [33]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [34]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

#drop the one NaN value from list and convert series to ints to be used for mapping
toronto_merged = toronto_merged.dropna()
toronto_merged["Cluster Labels"] = toronto_merged["Cluster Labels"].astype(int)

toronto_merged.head()

Postcode          Borough                   Neighbourhood   Latitude  \
0      M4E     East Toronto                     The Beaches  43.676845   
1      M4K     East Toronto    The Danforth West, Riverdale  43.683262   
2      M4L     East Toronto  The Beaches West, India Bazaar  43.667965   
3      M4M     East Toronto                 Studio District  43.662766   
4      M4N  Central Toronto                   Lawrence Park  43.728160   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.295225               0     Health Food Store                 Trail   
1 -79.355120               0              Bus Line                  Park   
2 -79.314673               0     Fish & Chips Shop          Liquor Store   
3 -79.334830               0    Italian Restaurant                Bakery   
4 -79.387085               4    Photography Studio              Bus Line   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub           Wings Joint  Ethiopian Restaurant   
1        Discount Store         Grocery Store           Wings Joint   
2      Sushi Restaurant         Movie Theater    Italian Restaurant   
3                 Diner                  Café           Pizza Place   
4           Swim School           Event Space                  Food   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                  Food           Flower Shop           Flea Market   
1           Event Space                  Food           Flower Shop   
2                   Pub          Intersection        Sandwich Place   
3               Brewery   American Restaurant   Arts & Crafts Store   
4           Flower Shop           Flea Market           Fish Market   

  9th Most Common Venue 10th Most Common Venue  
0           Fish Market      Fish & Chips Shop  
1           Flea Market            Fish Market  
2        Ice Cream Shop             Steakhouse  
3      Sushi Restaurant            Coffee Shop  
4     Fish & Chips Shop   Fast Food Restaurant

Visualize resulting clusters on Folium map centered on Toronto, color-coding each cluster

In [35]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' -- Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters